### PR TITLE
Leverage LazyGhostTrait when possible

### PR DESCRIPTION
**Pull request by nicolas-grekas**
This PR replaces Doctrine's proxies by [VarExporter's `LazyGhostTrait`](https://symfony.com/blog/revisiting-lazy-loading-proxies-in-php).

~It needs https://github.com/doctrine/persistence/pull/307~ (merged)

Requires running `composer require symfony/var-exporter:^6.2@dev doctrine/persistence:^3.1@dev` to enable the new proxies for now.

### CODE REVIEW

File: lib/Doctrine/ORM/Proxy/ProxyFactory.php [hunk 1]
```diff
@@ -55,7 +80,15 @@ public function __construct(EntityManagerInterface $em, $proxyDir, $proxyNs, $au
     {
         $proxyGenerator = new ProxyGenerator($proxyDir, $proxyNs);
 
-        $proxyGenerator->setPlaceholder('baseProxyInterface', Proxy::class);
+        if (trait_exists(LazyGhostTrait::class)) {
```
Comment (nicolas-grekas) [inline lib/Doctrine/ORM/Proxy/ProxyFactory.php]: Would need to be replaced by an explicit option to opt-in.

File: lib/Doctrine/ORM/Proxy/ProxyFactory.php [hunk 2]
```diff
@@ -142,16 +183,44 @@ private function createInitializer(ClassMetadata $classMetadata, EntityPersister
...
+            }
+
+            if ($entity === null) {
+                throw EntityNotFoundException::fromClassNameAndIdentifier(
+                    $classMetadata->getName(),
+                    $this->identifierFlattener->flattenIdentifier($classMetadata, $identifier)
+                );
+            }
+
+            Hydrator::hydrate($proxy, (array) $entity);
```
Comment (nicolas-grekas) [inline lib/Doctrine/ORM/Proxy/ProxyFactory.php]: Happens when a proxy is first detached then lazy-init is triggered.

File: lib/Doctrine/ORM/Proxy/ProxyFactory.php [hunk 3]
```diff
@@ -180,4 +249,57 @@ private function createCloner(ClassMetadata $classMetadata, EntityPersister $ent
             }
         };
     }
+
+    private function generateProxyModifiers(ClassMetadata $class): string
+    {
+        return PHP_VERSION_ID >= 80200 && $class->getReflectionClass()->isReadOnly() ? 'readonly ' : '';
```
Comment (nicolas-grekas) [inline lib/Doctrine/ORM/Proxy/ProxyFactory.php]: We could decide to reject readonly classes instead, until https://externals.io/message/118554 is resolved at least.

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 4]
```diff
@@ -2870,23 +2864,25 @@ public function createEntity($className, array $data, &$hints = [])
                             break;
 
                         default:
+                            $normalizedAssociatedId = $this->normalizeIdentifier($targetClass, $associatedId);
```
Comment (nicolas-grekas) [inline lib/Doctrine/ORM/UnitOfWork.php:2874]: Without normalization, proxies are initialized with type-incompatible values for the identifier fields!?!

File: tests/Doctrine/Tests/ORM/Functional/MergeProxiesTest.php [hunk 5]
```diff
@@ -67,7 +67,7 @@ public function testMergeUnserializedUnInitializedProxy(): void
 
         self::assertSame(
             $managed,
-            $this->_em->merge(unserialize(serialize($this->_em->merge($detachedUninitialized))))
+            $this->_em->merge($this->_em->merge($detachedUninitialized))
```
Comment (nicolas-grekas) [inline tests/Doctrine/Tests/ORM/Functional/MergeProxiesTest.php]: Proxies are initialized on serialize.

Comment (stof) [reply tests/Doctrine/Tests/ORM/Functional/MergeProxiesTest.php]: this kind of defeats the test IMO, as there is no unserialization involved anymore

File: composer.json [hunk 6]
```diff
@@ -36,7 +36,9 @@
         "psr/cache": "^1 || ^2 || ^3",
         "symfony/console": "^3.0 || ^4.0 || ^5.0 || ^6.0",
         "symfony/polyfill-php72": "^1.23",
-        "symfony/polyfill-php80": "^1.16"
+        "symfony/polyfill-php80": "^1.16",
+        "symfony/var-dumper": "^6.1",
```
Comment (stof) [inline composer.json]: why is it a required dependency now ?

Comment (nicolas-grekas) [reply composer.json]: for me, for debugging :)

**Comment by stof** [comment]
> As hinted in the description of the PR, Doctrine relies on the fact that setting a property would initialize a proxy. This forces a round-trip to the DB to fetch data that could be updated directly instead. It be great to figure out if this requirement really makes sense, and ditch it of not.

This requirement is a must-have. Without it, change tracking would be broken as Doctrine would not have the old value initialized to compare it on flush.

**Comment by stof** [comment]
or at least, it is a must-have with the current implementation of the UnitOfWork.

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 4 above)
Comment (stof) [reply lib/Doctrine/ORM/UnitOfWork.php:2874]: this looks like a bug in the existing Doctrine implementation too. I suggest submitting this fix against 2.13.x

**Comment by nicolas-grekas** [comment]
PR updated. Now that LazyGhostTrait initializes also when setting a property, https://github.com/doctrine/persistence/pull/307 is needed, and more importantly it's way easier to integrate with the ORM.

This PR is green locally. Ready for more extensive review.

@beberlei before I open this on the official repo of doctrine/orm, it'd be happy to know if that work is of any interest to Doctrine. I spent many weeks on it so I hope yes, but you're the one that's going to decide in the end :)

/cc @greg0ire, @derrabus and @malarzm also if you can have a look and maybe share your opinion on the PR?

File: lib/Doctrine/ORM/Proxy/ProxyFactory.php [hunk 7]
```diff
@@ -142,16 +199,46 @@ private function createInitializer(ClassMetadata $classMetadata, EntityPersister
...
+     *
+     * @psalm-return Closure(Proxy):void
+     *
+     * @throws EntityNotFoundException
+     */
+    private function createLazyInitializer(ClassMetadata $classMetadata, EntityPersister $entityPersister): Closure
+    {
+        return function (Proxy $proxy) use ($entityPersister, $classMetadata): void {
+            if ($classMetadata->getReflectionClass()->hasMethod('__wakeup')) {
+                $proxy->__wakeup();
```
Comment (nicolas-grekas) [inline lib/Doctrine/ORM/Proxy/ProxyFactory.php]: This is kept only for BC and is not needed from a technical pov.

File: docs/en/reference/architecture.rst [hunk 8]
```diff
@@ -159,17 +142,13 @@ Serializing entities
...
-parent classes. On the other hand it is not a solution for proxy
-objects to implement ``Serializable`` because Serializable does not
-work well with any potential cyclic object references (at least we
-did not find a way yet, if you did, please contact us). The
-``Serializable`` interface is also deprecated beginning with PHP 8.1.
+references to proxy objects or is still managed by an EntityManager.
+By default, serializing proxy objects does not initialize them. On
+unserialization, resulting objects are detached from the entity
+manager and cannot be initialiazed anymore. You can implement the
+``__unserialize()`` method if you want to change that behavior, but
```
Comment (stof) [inline docs/en/reference/architecture.rst]: this should be ``__serialize`` instead, which is the one allowing to trigger initialization during serialization

File: lib/Doctrine/ORM/Internal/Hydration/SimpleObjectHydrator.php [hunk 9]
```diff
@@ -148,10 +148,6 @@ protected function hydrateRowData(array $row, array &$result)
             }
         }
 
-        if (isset($this->_hints[Query::HINT_REFRESH_ENTITY])) {
-            $this->registerManaged($this->class, $this->_hints[Query::HINT_REFRESH_ENTITY], $data);
```
Comment (stof) [inline lib/Doctrine/ORM/Internal/Hydration/SimpleObjectHydrator.php:152]: why removing this ?

**Comment by greg0ire** [comment]
> share your opinion on the PR

The changes in the `docs` directory are the ones I can understand, and it appears you are lifting a good number of restrictions. Besides, I'm really looking forward to getting rid of `doctrine/common`, so big 👍 from me :slightly_smiling_face:   Regarding the technical side of things, I'll defer to Benjamin and Alexander, as I'm really unfamiliar with the proxy stuff.

File: lib/Doctrine/ORM/Internal/Hydration/SimpleObjectHydrator.php (hunk 9 above)
Comment (nicolas-grekas) [reply lib/Doctrine/ORM/Internal/Hydration/SimpleObjectHydrator.php:152]: I borrowed this from https://github.com/doctrine/orm/pull/6719/files#diff-c9c1faa5750ed5a13ec97a922d181d5b26a9f74fd79345cef5264725054cef2b and it makes tests pass without breaking any when using `Common\Proxy` but I must admit I didn't understand this completely yet. Anyone does?

**Comment by nicolas-grekas** [comment]
Only one failing test remains: `ENABLE_SECOND_LEVEL_CACHE=1 ./vendor/bin/phpunit --filter testCloneProxy`

The reason of the failure is that when the cloned proxy is initialized, it registers itself in the manager. The previous proxies were initialized specially when cloning. I'm not sure how to adjust the code for the new proxies.

**Pull request closed without merging**